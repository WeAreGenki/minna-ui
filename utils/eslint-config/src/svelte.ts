/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

interface BlockAttributes {
  type: string;
}

export = {
  plugins: ['svelte3'],
  settings: {
    'svelte3/ignore-styles': (attr: BlockAttributes) =>
      attr.type === 'text/postcss',
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
