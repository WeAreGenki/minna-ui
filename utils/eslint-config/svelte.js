/**
 * Minna UI Svelte & Sapper ESLint config add-on.
 *
 * @fileoverview Provides extra config for Svelte & Sapper projects. This
 * config is and add-on which should extend the base config preset.
 * @see https://eslint.org/docs/user-guide/configuring
 */

/* eslint-disable sort-keys */

'use strict';

/** @type {import('./types').ESLintConfig} */
module.exports = {
  plugins: ['svelte3'],
  settings: {
    /**
     * @param {import('./types').BlockAttributes} attr - Tag attributes.
     * @returns {boolean} If style block should be ignored.
     */
    'svelte3/ignore-styles': (attr) => attr.type === 'text/postcss',
    // TODO: Uncomment once false positives are handled better
    // 'svelte3/lint-template': true,
  },
  overrides: [
    // Svelte components
    {
      files: ['*.svelte'],
      rules: {
        // import order cannot be determined correctly in .svelte components
        'import/first': 'off',
        // components are compiled at build-time so devDependencies are OK
        'import/no-extraneous-dependencies': [
          'error',
          { devDependencies: true },
        ],
        // Svelte abuses the syntax with `export let`
        'import/no-mutable-exports': 'off',
      },
    },
  ],
};
