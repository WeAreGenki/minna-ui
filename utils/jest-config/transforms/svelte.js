/**
 * Jest transform for compiling Svelte components into JavaScript.
 *
 * NOTE: `svelte` is not listed as a dependency so developers can use this
 * jest preset for non-svelte projects.
 */

'use strict';

const { compile } = require('svelte/compiler'); // eslint-disable-line import/no-extraneous-dependencies

/** @type {import('../types').JestTransformer} */
exports.process = (src, filename) => {
  // Strip out <style> tags to prevent errors when unable to parse PostCSS etc.
  const re = /<style[^>]*>[\S\s]*?<\/style>/g;
  const normalised = src.replace(re, '');

  const result = compile(normalised, {
    accessors: true, // Easy access to component props for tests
    css: false,
    dev: true,
    filename,
    format: 'cjs',
    preserveWhitespace: true,
  });

  const esInterop =
    'Object.defineProperty(exports, "__esModule", { value: true });';

  return {
    code: result.js.code + esInterop,
    map: result.js.map,
  };
};
