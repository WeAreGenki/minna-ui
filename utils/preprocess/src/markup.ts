/* eslint-disable no-restricted-syntax, security/detect-object-injection */

import { MarkupPreprocessor } from './types';

/**
 * Minna UI svelte markup preprocessor.
 * Reduces the whitespace in svelte components to a minimum.
 *
 * @param opts - User defined options.
 * @param opts.enabled - Enable preprocessor to transform HTML code.
 */
export const markup = ({ enabled = true } = {}): MarkupPreprocessor => ({
  content,
}) => {
  if (!enabled) return;

  let code = `${content}`;
  let count = 0;

  const tags = [
    ['<script', '</script>'],
    ['<style', '</style>'],
    ['<pre', '</pre>'],
  ];
  const blocks = new Map();

  // replace tag blocks with markers
  for (const tag of tags) {
    let start;

    // eslint-disable-next-line no-cond-assign
    while ((start = code.indexOf(tag[0])) !== -1) {
      const end = code.indexOf(tag[1], start) + tag[1].length;
      const inner = code.slice(start, end);
      const marker = `<___marker_${count}>`;
      blocks.set(marker, inner);

      code = code.substring(0, start) + marker + code.substring(end);

      count += 1;
    }
  }

  // remove HTML comments
  code = code.replace(/<!--.*?-->/gsu, '');

  // remove whitespace between tags
  code = code.replace(/>\s*?</gm, '><');

  // reduce multiple whitespace down to a single space
  code = code.replace(/\s{2,}/gm, ' ');

  // convert remaining whitespace characters into a space
  code = code.replace(/\s/gm, ' ');

  // remove surrounding whitespace
  code = code.trim();

  // restore tag blocks
  for (const marker of blocks.keys()) {
    code = code.replace(marker, blocks.get(marker));
  }

  return {
    code,
  };
};
