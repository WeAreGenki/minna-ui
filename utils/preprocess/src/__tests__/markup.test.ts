import fs from 'fs';
import { promisify } from 'util';
// @ts-ignore - FIXME: Remove this line once the next version of Svelte is released
import { preprocess, create } from 'svelte/compiler';
import { markup } from '../markup';

// eslint-disable-next-line security/detect-non-literal-fs-filename
const readFile = promisify(fs.readFile);

// don't import component to avoid Jest transform
const componentPath = require.resolve('@minna-ui/jest-config/fixtures/TestComponent.svelte');

const opts = {
  markup: markup(),
  style: () => ({ code: '/*discarded*/' }),
};

const sourceSimple = `
<div id="target">
  test

  </div>

    <hr class="one two three">

  <input type="text" name="test" class ="" readonly="readonly">

      whitespace

  <!-- comment -->
  <!---->
`;
const sourceBadSyntax = '<div id="" class=">test</span>';
let source = '';

beforeAll(async () => {
  source = await readFile(componentPath, 'utf8');
});

describe('Svelte markup preprocessor', () => {
  it('processes without any error', () => {
    expect.assertions(2);
    const spy = jest.spyOn(process.stderr, 'write');
    function wrapper(): void {
      preprocess(sourceSimple, opts);
    }
    expect(wrapper).not.toThrow();
    expect(spy).not.toHaveBeenCalled();
    spy.mockRestore();
  });

  it.skip('processes a simple component', async () => {
    expect.assertions(4);
    const output = preprocess(sourceSimple, opts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).toMatch('> <');
    expect(result).not.toMatch('><');
    expect(result).toMatchSnapshot();
  });

  it('processes a component', async () => {
    expect.assertions(2);
    const output = preprocess(source, opts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).toMatchSnapshot();
  });

  it.skip('creates and mounts a component', async () => {
    expect.assertions(1);
    const processed = await preprocess(source, opts);
    const TestComponent = create(processed.toString());
    const target = document.createElement('div');
    new TestComponent({ target });
    expect(target.innerHTML).toMatchSnapshot();
  });

  it.skip('prints error on bad HTML syntax', async () => {
    expect.assertions(1);
    const spy = jest.spyOn(process.stderr, 'write');
    spy.mockImplementation(() => true);
    await preprocess(sourceBadSyntax, opts);
    expect(spy).toHaveBeenCalled();
    spy.mockRestore();
  });
});
