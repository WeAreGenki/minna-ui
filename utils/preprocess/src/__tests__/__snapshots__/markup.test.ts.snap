// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svelte markup preprocessor processes a component 1`] = `
"<script>
  export let name = 'Elon Musk';

  $: reversed = name.split('').reverse().join('');

  // refs
  let target;
  let nameReversed;
</script><style type=\\"text/postcss\\">/*discarded*/</style><div class=\\"wrapper\\"><div class=\\"text\\" bind:this=\\"{target}\\"> test {name} </div><div class=\\"text\\" bind:this=\\"{nameReversed}\\"> test <span>{reversed}</span></div></div>"
`;
